cmd/snap-update-ns/change.go

```diff
@@ -53,26 +54,40 @@ func (c Change) String() string {
 	return fmt.Sprintf("%s (%s)", c.Action, c.Entry)
 }
 
-var (
-	sysMount   = syscall.Mount
-	sysUnmount = syscall.Unmount
-)
-
-const unmountNoFollow = 8
-
 // Perform executes the desired mount or unmount change using system calls.
 // Filesystems that depend on helper programs or multiple independent calls to
 // the kernel (--make-shared, for example) are unsupported.
 func (c *Change) Perform() error {
-	switch c.Action {
-	case Mount:
-		flags, err := mount.OptsToFlags(c.Entry.Options)
-		if err != nil {
+	if c.Action == Mount {
+		mode := os.FileMode(0755)
+		uid := 0
+		gid := 0
+		// Create target mount directory if needed.
+		if err := ensureMountPoint(c.Entry.Dir, mode, uid, gid); err != nil {
 			return err
 		}
-		return sysMount(c.Entry.Name, c.Entry.Dir, c.Entry.Type, uintptr(flags), "")
+		// If this is a bind mount then create the source directory as well.
+		// This allows snaps to share a subset of their data easily.
+		flags, _ := mount.OptsToCommonFlags(c.Entry.Options)
+		if flags&syscall.MS_BIND != 0 {
+			if err := ensureMountPoint(c.Entry.Name, mode, uid, gid); err != nil {
+				return err
+			}
+		}
+	}
+	return c.lowLevelPerform()
+}
+
+// lowLevelPerform is a version of Perform that doesn't contain any special behavior.
```

Comment:
They are separated because code further ahead needs to call lowLevelPerform. I'll adjust comments on this but I need them separate in one way or another.

Reply:
Done now